### PR TITLE
feat: add Docusaurus static generated documentation

### PR DESCRIPTION
**Pull request by 0-vortex**
## What type of PR is this? (check all applicable)

- [x] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] 👷 Optimization
- [x] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description

As previously discussed with @bdougie I have essentially tested (all) most static documentation generators that were suitable for [opensauced](https://opensauced.pizza) and ultimately focused development effort towards [docusaurus](https://docusaurus.io).

This PR eliminates most of the projected discussions in #1039 and brings some visual examples we can work with. Personally think it's a big change that would not fit into a single issue which is why I did not turn that discussion into a half-assed issue.

## Related Tickets & Documents

Fixes multiple things, TBD based on how wholesome this PR is.

## Mobile & Desktop Screenshots/Recordings

Screenshot 2021-07-03 at 03 09 09
Screenshot 2021-07-03 at 03 10 06
Screenshot 2021-07-03 at 03 10 16
Screenshot 2021-07-03 at 03 10 24

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 readme
- [x] 📜 contributing.md
- [x] 📓 docs
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any pre/post-deployment tasks we need to perform?

- replace some README.md links with docs.opensauced.pizza 
- [peaceiris/actions-gh-pages](https://github.com/peaceiris/actions-gh-pages) command fails the first time behind protected branch unless you had previously set up GitHub Pages
- remove some of the static links in markdown in favour of static assets
- reorganise categories and documentation articles
- setup algolia free tier or custom search
- add some plugins
- unify theming by importing colors.js into ifs theme
- unify asset folders (better than pre-build copy hook that would break local docs development)
- document how to add docusaurus documentation

### CODE REVIEW

File: docs/docusaurus.config.js
Comment (filiptronicek) [inline docs/docusaurus.config.js:19]:
```suggestion
        alt: 'Open Sauced Logo',
```

File: docs/docusaurus.config.js
Comment (filiptronicek) [inline docs/docusaurus.config.js:30]: The GitHub link should probably lead to our repo page

```suggestion
          href: 'https://github.com/open-sauced/open-sauced',
```

**Review by filiptronicek — CHANGES_REQUESTED** [review]
Love the docs, good job, @0-vortex, this looks absolutely awesome!

I have found a couple of config things so far, and ESLint is bugging me with some issues about double quotes mainly.

**Comment by 0-vortex** [comment]
> Love the docs, good job, @0-vortex, this looks absolutely awesome!
> 
> I have found a couple of config things so far, and ESLint is bugging me with some issues about double quotes.

Give me more details about this, I removed `docs/` from being linted in https://github.com/open-sauced/open-sauced/pull/1078/commits/ab19d8c40ee253fc50b23af393856e30a279ccad but it might be possible IDEs rewrite that somehow and account for it :D

**Comment by filiptronicek** [comment]
> 
> 
> > Love the docs, good job, @0-vortex, this looks absolutely awesome!
> > I have found a couple of config things so far, and ESLint is bugging me with some issues about double quotes.
> 
> Give me more details about this, I removed `docs/` from being linted in [ab19d8c](https://github.com/open-sauced/open-sauced/commit/ab19d8c40ee253fc50b23af393856e30a279ccad) but it might be possible IDEs rewrite that somehow and account for it :D

Oh, it's probably just inheriting the config from the base repo.

This brings me to the question of: is it worth creating the docs in this repo? I think it deserves its own one, it will be easier to maintain and update (dependabot would have to be reconfigured), every change won't have to be deployed to both deploys. This is a big change, but I am all for it! What do you think, @0-vortex, @bdougie?

**Review by bdougie — COMMENTED** [review]
Left a few comments. I would prefer to use Netlify over GitHub Pages.

**Comment by bdougie** [comment]
> This brings me to the question of: is it worth creating the docs in this repo? I think it deserves its own one, it will be easier to maintain and update (dependabot would have to be reconfigured), every change won't have to be deployed to both deploys. This is a big change, but I am all for it! What do you think, @0-vortex, @bdougie?

IMO, when docs are in the same repo, they get noticed. I think this could be split out to a separate repo eventually but at the moment i like encouraging contributors to add docs in the same PR.

File: .github/workflows/ci.yml
Comment (0-vortex) [reply .github/workflows/ci.yml:64]: this is an artefact from actually testing the action in my fork - did not know what you had in mind for infrastructure at the time and ended up force pushing upstream by accident

File: docs/docs/contributing/code-of-conduct.md
Comment (0-vortex) [reply docs/docs/contributing/code-of-conduct.md:1]: Was thinking about how that file makes sense in both places, still think it's worth to somehow mirror their content rather than just link one to another. 

Think that's doable without chores? Judging this by the number of updates that file gets and some negative experience running parallel build systems (copy stuff from one place to another as a npm pre or git hook)

**Comment by 0-vortex** [comment]
> 
> > This brings me to the question of: is it worth creating the docs in this repo? I think it deserves its own one, it will be easier to maintain and update (dependabot would have to be reconfigured), every change won't have to be deployed to both deploys. This is a big change, but I am all for it! What do you think, @0-vortex, @bdougie?
> 
> IMO, when docs are in the same repo, they get noticed. I think this could be split out to a separate repo eventually but at the moment i like encouraging contributors to add docs in the same PR.
> 
> Adding some infrastructure to the deploys will hopefully make this an after thought. I have been asking around for ideas from other maintainers.
> 
> 

Coming from a node microservices background, I don't think code splits are painful if we leverage the same practices we do in the root folder. 

For docusaurus and docz I noticed our dependency tree becoming too complex and judging by previous discussions around making parts of open-sauced reusable modules, starting that effort with the docs endpoint seems worthwhile only if we want to test some future patterns, the docs themselves being crucial in visual content distribution and centralising our different historical issues.

That is also part of the reason I put so many discussion topics in the pull request (mostly unrelated to infrastructure) - there is a lot of value in consolidating some previous discussions to static docs, linking them to existing commits and otherwise retraceable information.

Gladly welcome the docs endpoint becoming a knowledge base worth additional maintenance 🔥

**Comment by mtfoley** [comment]
Not a whole lot to add on requested review. I like the screenshots, lol 👍  RE: deploying docusaurus to netlify, might be cool to do onstream with a pair programming deal. 

My bandwidth is wicked low lately with the day job, but putting this on netlify is something I'd love to watch on a stream.

**Review by mtfoley — APPROVED** [review]
LGTM, but then again, so do most things I haven't done before ;-)

**Review by filiptronicek — CHANGES_REQUESTED** [review]
From this second review, everything seems fine, just a tiny proposal: the colors. I think the default ones are nice, but some Open Sauced colors would fit the docs quite nicely. 

In the docs of a very nice React Native library, I found that we can do that in the custom CSS a little bit like so:
https://github.com/gorhom/react-native-bottom-sheet/blob/71993f0881feca9f70e010c75972fafb7c727c3a/website/src/css/custom.css#L5-L38

**Comment by 0-vortex** [comment]
> From this second review, everything seems fine, just a tiny proposal: the colors. I think the default ones are nice, but some Open Sauced colors would fit the docs quite nicely.
> 
> In the docs of a very nice React Native library, I found that we can do that in the custom CSS a little bit like so:
> https://github.com/gorhom/react-native-bottom-sheet/blob/71993f0881feca9f70e010c75972fafb7c727c3a/website/src/css/custom.css#L5-L38

Mentioned that previously, mostly because I tried to do it before opening the pull request and couldn't handle the conversion to infima. Personal judgement but while the colours were more in line with open sauced, there were screens that were pure ugly without extending it a bit more with CSS. 

> unify theming by importing colors.js into ifs theme

TL;DR: So disappointed by my colours.js implementation I reverted to defaults :D 

Will try with the colours you linked, but in definite need of help here :D

File: .github/workflows/ci.yml
Comment (bdougie) [reply .github/workflows/ci.yml:64]: Actually TBH, using the gh-pages branch works just as well.

**Review by bdougie — APPROVED** [review]
LGTM. Looking forward to iterating on this.

**Review by filiptronicek — APPROVED** [review]
I think I'm fine with the colors then, ready to merge! ✅

**Comment by 0-vortex** [comment]
> I think I'm fine with the colors then, ready to merge! ✅

On hold until #1034 is investigated, also adds some docs

**Comment by 0-vortex** [comment]
@filiptronicek @bdougie it's ready now 🍕

**Pull request merged**
